### DISCUSSION TITLE
Cleaner, more robust and efficient error handling for grofer

### DISCUSSION THREAD

**Discussion by MadhavJivrajani** [pesos/grofer#94] [Ideas]
Related issue: #92 
This might be a good starting point: https://github.com/cockroachdb/errors

**Comment by anihm136** [comment]
Okay I'm not very familiar with the overall architecture, but this is what I envision - a single thread/function as an error handler, which is capable of handling all errors gracefully. When any of the individual thread hit an error, pass the error (and control) to the error handling thread.
This is a bit like ISRs at a high level. Not sure how efficient or viable it is, but it would make future handling of errors easier

**Comment by Gituser143** [comment]
![image](https://user-images.githubusercontent.com/44526455/106493963-d5a86680-64df-11eb-961c-15216156a039.png)

Here's an image from Go Systems conf, SF 2020. We could implement this in places where there are a bunch of consecutive error checks, like in `updateProcess.go`

**Reply by anihm136** [reply]
Yeah, the example that is given later on is the exact thing I was thinking of. Hit an error, transfer control to a handler thread with relevant information

**Reply by Gituser143** [reply]
@anihm136, we've already got a bunch of go routines, I wouldn't want to introduce another solely for error handling unless we know for a fact it improves performance in some way. :grimacing: 

PS: I say this because although we already have error handling implemented (to some extent), we thought maybe we could cut grofer's resource utilisation with a more efficient form of error handling like mentioned in the video.

**Reply by anihm136** [reply]
Yeah routine was probably the wrong thing. What I meant was something similar to the method shown in that talk, where all functions panic and a handler is deferred inside each of them. The handler then calls a central function (or maybe one of a group from an errors package if necessary) which does the right thing

**Reply by anihm136** [reply]
Okay I'm seeing one major issue here - almost all existing packages return errors instead of panicking. So what happens is that you can ignore the error and return the value as-is, and it will cause panic somewhere else, which is hard to predict and messy to handle. On the other hand, any attempt to manually panic on error will still need to perform comparison, defeating the purpose of the excercise. Any ideas?
